frontend/lib/src/components/widgets/AudioInput/ActionButton.tsx

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from "react"
+
+import { useTheme } from "@emotion/react"
+import { Mic } from "@emotion-icons/material-outlined"
+import { Pause, PlayArrow, StopCircle } from "@emotion-icons/material-rounded"
+import RecordPlugin from "wavesurfer.js/dist/plugins/record"
+import WaveSurfer from "wavesurfer.js"
+
+import BaseButton, {
+  BaseButtonKind,
+} from "@streamlit/lib/src/components/shared/BaseButton"
+import Icon from "@streamlit/lib/src/components/shared/Icon"
+
+interface ActionButtonProps {
+  recordPlugin: RecordPlugin | null
+  recordingUrl: string | null
+  wavesurfer: WaveSurfer | null
+  hasNoMicPermissions: boolean
+  disabled: boolean
+  startRecording(): void
+  stopRecording(): void
+  onClickPlayPause(): void
+}
+
+const ActionButton: React.FC<ActionButtonProps> = ({
+  recordPlugin,
+  recordingUrl,
+  wavesurfer,
+  hasNoMicPermissions,
+  disabled,
+  startRecording,
+  stopRecording,
+  onClickPlayPause,
+}) => {
+  const theme = useTheme()
+
+  if (recordPlugin?.isRecording()) {
+    // It's currently recording, so show the stop recording button
+    return (
+      <BaseButton
+        kind={BaseButtonKind.BORDERLESS_ICON}
+        onClick={stopRecording}
+        disabled={disabled}
+        aria-label="Stop recording"
+      >
+        <Icon content={StopCircle} size="lg" color={theme.colors.primary} />
+      </BaseButton>
```

Comment:
I am not sure whether it would make it better or worse, but one thought I had here was to move this block to a function and re-use it like

```typescript
const createActionButton = (onClick, disabled, ariaLabel, iconContent, color) => {
  return (
    <BaseButton
        kind={BaseButtonKind.BORDERLESS_ICON}
        onClick={onClick}
        disabled={disabled}
        aria-label={ariaLabel}
      >
        <Icon content={iconContent} size="lg" color={color} />
      </BaseButton>
  )
}

if (recordPlugin?.isRecording()) {
  return createActionButton(stopRecording, disabled, "Stop recording", StopCircle, theme.colors.primary)
} 
else if (recordingUrl) {
  if (wavesurfer && wavesurfer.isPlaying()) {
    return createActionButton(onClickPlayPause, disabled, "Pause", Pause, theme.colors.fadedText60)  
  }

...
} 
```

It might make it easier to see on one glimpse that always the same Components (`<BaseButton><Icon/></BaseButton>`) are rendered.

But no strong opinion, just wanted to share my thought process 🙂

Reply:
if i have the time, happy to make this change